Fix: permissions validation regression

jakubcolony opened the pull request:
## Description

This issue has had its scope changed twice so it's possible that this PR contains some unrelated stuff, but the main point is to re-introduce the validation logic in a new form as outlined in the issue.

As usual, there is a fair bit of refactoring and moving stuff around as well.

Please note point 4 from the issue is not included in this PR:
> Non-connected users (may need a separate issue)
> Do same as "Users with no permissions at all for the action" but instead show non-logged in message.
> Message: "You are not signed in, please connect your wallet to create actions."
> https://www.figma.com/file/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?type=design&node-id=20469-232289&mode=design&t=hJ7bZDuFrLX6bk0u-4

@arrenv I believe this can only be done after the colony gating is over? Right now you can't access the action sidebar without being logged in.

## Testing

For testing, please refer to the original issue.

Resolves #1778 
Resolves #1866

iamsamgibbs reviewed (changes requested):
This is a great start. The new permissions helpers and hooks are very nice.

I’ve noticed just a couple of things.

From the original ticket, when the reputation weighted extension is enabled: `Users with no permissions at all can see only the Reputation decision method.`

In this screenshot, the user has no permissions at all but can still see both decision methods:

![Screenshot 2024-02-08 at 10 32 46](https://github.com/JoinColony/colonyCDapp/assets/34915414/ab912884-6adc-454d-b715-b0d4105e3687)

The “You don’t have the right permissions to create this action” error message shows when viewing already completed actions.

![Screenshot 2024-02-08 at 10 16 48](https://github.com/JoinColony/colonyCDapp/assets/34915414/73233fb1-708b-4d34-a7b8-85b4b2c4d14f)

(Might be out of the scope of the ticket, but the fields on the “Manage colony objective” action don’t grey out when fields are disabled.)

![Screenshot 2024-02-08 at 10 14 34](https://github.com/JoinColony/colonyCDapp/assets/34915414/e3859330-9f10-4b47-b2fe-e721450a4e94)

ArmandoGraterol commented on src/components/v5/common/ActionSidebar/hooks/permissions/helpers.ts:
```diff
@@ -0,0 +1,101 @@
+  switch (actionType) {
+    case ACTION.SIMPLE_PAYMENT:
+      return [ColonyRole.Funding, ColonyRole.Administration];
+    case ACTION.MINT_TOKENS:
+      return [ColonyRole.Root];
+    case ACTION.TRANSFER_FUNDS:
+      return [ColonyRole.Funding];
+    case ACTION.UNLOCK_TOKEN:
+      return [ColonyRole.Root];
+    case ACTION.MANAGE_TOKENS:
+      return [ColonyRole.Root];
+    case ACTION.CREATE_NEW_TEAM:
+      return [ColonyRole.Architecture];
+    case ACTION.EDIT_EXISTING_TEAM:
+      return [ColonyRole.Architecture];
+    case ACTION.MANAGE_REPUTATION:
+      /**
+       * @TODO: Once this action is wired, we'll need to tell if
+       * it's a smite or award action (most likely from `formValues`)
+       * If smite: Arbitration, else: Root
+       */
+      return undefined;
+    case ACTION.MANAGE_PERMISSIONS: {
+      if (!formValues.team) {
+        return undefined;
+      }
+      return formValues.team === Id.RootDomain
+        ? [ColonyRole.Root, ColonyRole.Architecture]
+        : [ColonyRole.Architecture];
+    }
+    case ACTION.EDIT_COLONY_DETAILS:
+    case ACTION.MANAGE_COLONY_OBJECTIVES:
+      return [ColonyRole.Root];
+    case ACTION.UPGRADE_COLONY_VERSION:
+      return [ColonyRole.Root];
+    case ACTION.ENTER_RECOVERY_MODE:
+      return [ColonyRole.Recovery];
```
We could group all of the cases that use `Root` here.
```
case ACTION.MINT_TOKENS:
case ACTION.UNLOCK_TOKEN:
case ACTION.MANAGE_TOKENS:
case ACTION.EDIT_COLONY_DETAILS:
case ACTION.MANAGE_COLONY_OBJECTIVES:
case ACTION.UPGRADE_COLONY_VERSION:
  return [ColonyRole.Root];
```

ArmandoGraterol replied:
@jakubcolony Could it be that perhaps you missed this suggestion? :)

ArmandoGraterol reviewed (changes requested):
Looking good comrade Jakub! I've noticed what Sam already pointed out but besides that, I noticed this is also not happening:

>Users with at least one permission in any team related to the action
>- If the team has been changed, then it should reset the Decision method field.


EDIT: Also this small detail:

![Screenshot 2024-02-08 at 10 46 38](https://github.com/JoinColony/colonyCDapp/assets/18473896/5452bfe3-0537-488a-ab1e-87219ce6a8ae)

The colony description is greyed out when disabled but the name isn't. I think the description shouldn't be, no? At least that's what I infer from looking at how the other disabled prefilled fields look.

jakubcolony commented:
Thank you for the reviews. I believe I fixed all the issues reported apart from:

> If the team has been changed, then it should reset the Decision method field.

I think we should drop this requirement as it doesn't impact validation but would require yet another layer of knowledge as to which fields are needed to determine required permissions (changing any of them, not just team, would require decision method reset to be consistent).

I will also not fix the non-grayed out fields in this PR, as it's an issue with the `InputBase` component not showing disabled state correctly - I added a note to #1809 to fix that.

iamsamgibbs reviewed (changes requested):
Thanks for making those changes, previous issues all appear resolved now.

I've only noticed one new issue. With a user who has only "payer" permissions, select the "Manage permissions" action. The invalid permissions error message should appear at the top straight away (as it does with the other actions), but instead it only appears once you select a team.

![Screenshot 2024-02-09 at 09 07 51](https://github.com/JoinColony/colonyCDapp/assets/34915414/3faf5b5b-1f55-4714-9b00-2e330e39bd37)

![Screenshot 2024-02-09 at 09 08 06](https://github.com/JoinColony/colonyCDapp/assets/34915414/f6e3d7c2-f3c8-48f8-b56f-a0b5aba53e5c)

jakubcolony commented:
@iamsamgibbs great spot, I made changes to how `useHasNoDecisionMethods` work. Ready for a re-test!

rdig commented on src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/partials/NoReputationError.tsx:
```diff
@@ -0,0 +1,52 @@
+        icon="warning-circle"
+        description={formatMessage(MSG.noReputationError)}
+        callToAction={
+          <Link to="https://docs.colony.io/use/reputation">
```
This needs to be URL constant

jakubcolony replied:
These aren't my changes and I want to avoid bloating this PR any more with unrelated fixes

rdig replied:
In that case, just make sure these changes aren't lost, and we keep track of them

rdig commented on src/components/v5/common/ActionSidebar/partials/ColonyDetailsFields/ColonyDetailsFields.tsx:
```diff
@@ -9,82 +9,94 @@ import ActionFormRow from '~v5/common/ActionFormRow/index.ts';
+        <div className="flex items-center">
+          <ColonyAvatarField
+            name="avatar"
+            fileOptions={{
+              fileFormat: ['.PNG', '.JPG', '.SVG'],
+              fileDimension: '250x250px',
+              fileSize: '1MB',
+            }}
```
This should be some sort of constant as well, as I feel weird having them declared specifically in the component.

rdig commented on src/components/v5/common/ActionSidebar/partials/forms/BatchPaymentForm/BatchPaymentForm.tsx:
```diff
@@ -32,27 +29,7 @@ const BatchPaymentForm: FC<ActionFormBaseProps> = () => {
-          title={formatText({ id: 'actionSidebar.availableDecisions' })}
-        />
-      </ActionFormRow>
+      <DecisionMethodField />
```
Wasn't the decision method component extracted in #1882 ? Or is this just a case of needing a rebase ?

jakubcolony replied:
That PR is still open, so I guess whoever merges first?

rdig reviewed (dismissed):
Left some minor comments that I picked up on, otherwise this is all good.

Nice work!

ArmandoGraterol reviewed (approved):
Looks good to me! Nice work Jakub 🎉

iamsamgibbs reviewed (changes requested):
The issue with the `manage permissions` is fixed now! But I've noticed a couple of other error validation issues when switching between action types.

https://github.com/JoinColony/colonyCDapp/assets/34915414/e1f7373a-f9f1-4899-87c4-cd5c88931de7

After I stopped recording, I switched to an action I do have permissions for, and the error validation messages still include fields from the previous action.

![Screenshot 2024-02-12 at 11 08 19](https://github.com/JoinColony/colonyCDapp/assets/34915414/165e6d47-b1d2-4922-89e1-25737c8c5af2)

arrenv reviewed (commented):
Great work on this @jakubcolony, other then the same issue as @iamsamgibbs, all the logic seems to work. I will test again when it is in a ready state though as there is a lot of validation happening all over the place.

jakubcolony commented:
@iamsamgibbs thanks for having another look, I went through the issues you highlighted and fixed some of the fields showing in red color while they're in both error and disabled states (disabled should take precedence). 

All the other issues such as incorrectly validating when switching action type and showing error messages seem to be present in master.

iamsamgibbs reviewed (dismissed):
Great job @jakubcolony - all seems resolved now 👍

rdig reviewed (approved):
Approving, nice work Jakub

arrenv reviewed (commented):
Great work @jakubcolony, all working great for every combination I could think of.

arrenv reviewed (approved):
Re-approving
